### DISCUSSION TITLE
Question: Should Proxy Authentication compliment Cookies Authentication?

### DISCUSSION THREAD

**Discussion by VladimirCores** [apache/couchdb#4693] [General]
With following config settings fro auth:
```
[couchdb]
users_db_security_editable = true

[chttpd]
authentication_handlers = {chttpd_auth, cookie_authentication_handler}, {chttpd_auth, default_authentication_handler}, {chttpd_auth, jwt_authentication_handler}, {chttpd_auth, proxy_authentication_handler}
allow_jsonp = false
require_valid_user = false
require_valid_user_except_for_up = false

[chttpd_auth]
users_db_public = false
secret = bf5734a87048aba5e1c5876bfa88e12a
proxy_use_secret = true
timeout = 6000
hash_algorithms = sha256, sha
```

Request to `_users` database only possible with default (Basic) or cookies authentication and  "Proxy" (headers with `X-Auth-CouchDB-Token/Roles/UserName`) has no effect whatsoever. 

How to enable Proxy Authentication to be an additional layer of security when `X-Auth-CouchDB-Token` is mandatory for request to succeed?

**Comment by VladimirCores** [comment]
I mean that GET request for the session, after POST to `_session` with user credentials, succeed (!) even without proxy headers:
![image](https://github.com/apache/couchdb/assets/442111/f6c2bd36-a222-4a91-b8d4-55355c707293)

Am I doing something wrong?

**Comment by VladimirCores** [comment]
And when requesting same user with Proxy auth it says:
![image](https://github.com/apache/couchdb/assets/442111/8a7336b7-cdb4-4231-ba7b-24627304a1e2)

Why? What is the reason for Proxy Authentication then? Or should it behave same as Cookies Auth and allow to return user info?

**Comment by big-r81** [comment]
Hi, here are more infos about [Proxy Authentication](https://docs.couchdb.org/en/latest/api/server/authn.html#proxy-authentication).

If not configured otherwise, only users of role `_admin` can access and edit documents in `_users`. Your user `geom` doesn't have this role.

**Reply by VladimirCores** [reply]
I have read the docs carefully and then disabled this option when set `[couchdb] users_db_security_editable = true` (on first attached image).
Also I put role `user` to the members of the `_users` database.

![image](https://github.com/apache/couchdb/assets/442111/88afdace-cf09-4782-ba60-29519bbc6a3f)

But still can't authorize with Proxy ... looks like something else, but what?

**Comment by big-r81** [comment]
First, check if your proxy auth is working (above you authenticated with cookie authentication)

It should look like this (`GET /_session`):
```json
{
    "ok": true,
    "userCtx": {
        "name": "proxy",
        "roles": [
            "PROXY-USER-ROLE1",
            "PROXY-USER-ROLE2"
        ]
    },
    "info": {
        "authentication_handlers": [
            "proxy",
            "cookie",
            "default"
        ],
        "authenticated": "proxy"
    }
}
```
The important part is `"authenticated": "proxy"`.

**Reply by VladimirCores** [reply]
Yeah, forgot to mention this - and that is the problem - Proxy Authentication is not working. And I dont understand why : |

![image](https://github.com/apache/couchdb/assets/442111/b9dfe91e-b05e-4f22-92fb-306669ce5bf4)

**Comment by rnewson** [comment]
The authentication handlers cannot be combined.

you seem to have configured things correctly and your request works for me when I apply the same settings;

```
curl localhost:15984/_session -H X-Auth-CouchDB-UserName:geom \
-H X-Auth-CouchDB-Token:f6cfd45419f6b5571f68af00588fba744d1957c97b3ccdd069ee530d060ea2c2

{"ok":true,"userCtx":{"name":"geom","roles":[]},"info":{"authentication_handlers": \
["cookie","default","jwt","proxy"],"authenticated":"proxy"}}
```

**Reply by VladimirCores** [reply]
And this is exactly the topic of the discussion - should Proxy be addition (&&) to Cookies?

As for me, it should, and I would make it as an additional layer of protection as I see Proxy being used in Bastion Host or reverse proxy to protect the data even more.

**Comment by rnewson** [comment]
For cookie authentication you need to know a secret (a password) stored in CouchDB. For proxy authentication you need to know a secret (the proxy secret) stored in CouchDB. There isn't any obvious greater protection in a request needing to know two secrets instead of one.

It doesn't make sense to me to require multiple authentication handlers for a request, at least not the two in this example.

**Reply by VladimirCores** [reply]
I did not find any mentions about priority of the Cookie authentication over Proxy. The setting of `proxy_use_secret = true` mean to me that the Proxy auth is enabled, but as you (@rnewson) wrote its not and this is confusing (!), also because both methods included in authentication_handlers list.

As for me, using them both opens advanced/additional way of app protection when this proxy token could be injected in the app (and requests from it) or in reverse-proxy.

Maybe documentation should be updated?

**Reply by rnewson** [reply]
The order of handlers in `authentication_handlers` is significant. Each handler is tried in order until one of them authenticates the request, or the end of the list is reached. In your config above the cookie handler is listed before the proxy handler.

**Reply by VladimirCores** [reply]
> The order of handlers in authentication_handlers is significant.

Its hard to guess this rule. Double checked the docs on [`authentication_handlers`](https://docs.couchdb.org/en/stable/config/http.html#chttpd/authentication_handlers) nothing about order of handlers : |

And why then I can't get user data (from `_users`) with Proxy auth of the same user?
see: https://github.com/apache/couchdb/discussions/4693#discussioncomment-6520218

**Reply by rnewson** [reply]
Hm, it might not be explicitly written anywhere, would you submit a documentation PR?

The reason you can't get user data from `_users` is that they are _not_ the same user. They just have the same name.

Again, it is confusing, and not intended, to use proxy authentication handler _and_ the users db. The proxy authentication handler exists specifically to allow administrators to use an external source of truth _instead of_ the CouchDB users database.

**Reply by VladimirCores** [reply]
1. About PR to documentation - it would be a honor for me to contribute to one of my favorite project CouchDB ; ) However I have to understand the process of authorization better since there are few things I don't know and need to try - How exactly Proxy authentication works, Cookies and JWT (I have tried JWT auth with external oauth server [authorizer.dev](https://authorizer.dev/) about half-year ago and it worked) but not together with Cookies (in same scenario as Proxy). 

2. 
> The proxy authentication handler exists specifically to allow administrators to use an external source of truth instead of the CouchDB users database.

And how exactly should Proxy auth be used? `_users` database has Role - `user`(see [here](https://github.com/apache/couchdb/discussions/4693#discussioncomment-6521254)), the `user` role  which Proxy's header includes, but still no effect, and I have tried same thing with other databases. Not clean how it works and what "user" should be encoded to proxy's token (with HMAC) like in [the example](https://docs.couchdb.org/en/stable/api/server/authn.html#proxy-authentication)?

**Comment by VladimirCores** [comment]
Back to the topic of Proxy Authentication in CouchDB @rnewson - what are these "proxy" users and how can I manage them in CouchDB?

**Comment by rnewson** [comment]
The sole point of the proxy handler is that users and roles are controlled externally (by something that knows the shared secret or is proxying all requests to couchdb to add the x-auth headers), so CouchDB doesn't know the users or roles and does not manage them by design.
